### DISCUSSION TITLE
Suggestion: Setting up a truly transparent proxy.

### DISCUSSION THREAD

**Discussion by H0mTanks** [emanuele-em/proxelar#38] [Q&A]
Right now, the user has to manually set up a proxy in their OS settings and set up their browser to use the OS proxy. This means the user has to make sure that whatever application they use is using the OS's proxy settings.
My use case involves a proxy being truly transparent, in the sense that a network driver can be used to route all the traffic to the proxy. 
This can most likely be done by writing an eBPF network driver for both Windows and Linux (dunno about OSX). This eBPF driver would hook into `cgroup/connect4`, `cgroup/connect6`, `cgroup/recv_accept4`, `cgroup/recv_accept6` and redirect all connection packets to the proxy on a driver level by intercepting `connect()` and `accept()` syscalls.

**Comment by emanuele-em** [comment]
Yes sure! This would be a **great** upgrade.
This solution seems to be a good idea also if I'm not a big expert in field.
1. @H0mTanks is this allow users to intercept also traffic from the same machine that launch the proxy?
2. How can it distinguish proxy traffic from other processes traffic to avoid loops? (iptables and pid?)

**Reply by H0mTanks** [reply]
1. Yes, intercepting traffic by running as an app on the same machine that launched it is my primary use case. By using a network driver, We can route the traffic to any IP Address and port that we want.

   The way the network driver would work is by intercepting the connection at the Transport Layer in the kernel and rewriting the Destination IP and the Destination Port of the packet, thereby being redirected to the proxy.

    A problem I was facing in this approach was that since the destination IP and Port get overwritten, I'm unsure how the proxy would then pass on the packets to the server? This is why I wanted to test against this project since the passthrough code is being handled for me, however I cannot test because the project does not run in a VM, hence the network driver work is blocked. See #39.

2. Could you elaborate further? I'm not very familiar with linux networking tools.

**Reply by emanuele-em** [reply]
ok, it's quite clear, the issue is correlated with egui interface, to test it you could bypass egui runnnig only the API crate ([proxyapi](https://github.com/emanuele-em/man-in-the-middle-proxy/tree/master/proxyapi)), anyway the problem seems to be common through egui community, I sent you one related [egui issue](https://github.com/emilk/egui/issues/2545).
